src/polynomials/Polynomial.jl

```diff
@@ -0,0 +1,187 @@
+    coeffs::Vector{T}
+    var::Symbol
+    function Polynomial{T}(coeffs::Vector{T}, var::Symbol) where {T <: Number}
+        length(coeffs) == 0 && return new{T}(zeros(T, 1), var)
+        last_nz = findlast(!iszero, coeffs)
+        last = max(1, last_nz === nothing ? 0 : last_nz)
+        return new{T}(coeffs[1:last], var)
+    end
+end
+
+@register Polynomial
+
+domain(::Type{<:Polynomial}) = Interval(-Inf, Inf)
+mapdomain(::Type{<:Polynomial}, x::AbstractArray) = x
+
+"""
+    (p::Polynomial)(x)
+
+Evaluate the polynomial using [Horner's Method](https://en.wikipedia.org/wiki/Horner%27s_method), also known as synthetic division.
+
+# Examples
+```jldoctest
+julia> p = Polynomial([1, 0, 3])
+Polynomial(1 + 3*x^2)
+
+julia> p(0)
+1
+
+julia> p.(0:3)
+4-element Array{Int64,1}:
+  1
+  4
+ 13
+ 28
+```
+"""
+function (p::Polynomial{T})(x::S) where {T,S}
+    R = promote_type(T, S)
+    length(p) == 0 && return zero(R)
+    b = convert(R, p[end])
+    @inbounds for i in (lastindex(p) - 1):-1:0
+        b = p[i] + x * b
+    end
+    return b
+end
+
+function fromroots(P::Type{<:Polynomial}, r::AbstractVector{T}; var::SymbolLike = :x) where {T <: Number}
+    n = length(r)
+    c = zeros(T, n + 1)
+    c[1] = one(T)
+    for j in 1:n, i in j:-1:1
+        c[i + 1] = c[i + 1] - r[j] * c[i]
+    end
+    return Polynomial(reverse(c), var)
+end
+
+function vander(P::Type{<:Polynomial}, x::AbstractVector{T}, n::Integer) where {T <: Number}
+    A = Matrix{T}(undef, length(x), n + 1)
+    A[:, 1] .= one(T)
+    @inbounds for i in 1:n
+        A[:, i + 1] = A[:, i] .* x
+    end
+    return A
+end
+
+function integrate(p::Polynomial{T}, k::S) where {T,S <: Number}
+    R = promote_type(eltype(one(T) / 1), S)
+    if hasnan(p) || isnan(k)
+        return Polynomial([NaN])
+    end
+    n = length(p)
+    a2 = Vector{R}(undef, n + 1)
+    a2[1] = k
+    @inbounds for i in 1:n
+        a2[i + 1] = p[i - 1] / i
+    end
+    return Polynomial(a2, p.var)
+end
+
+function derivative(p::Polynomial{T}, order::Integer = 1) where {T}
+    order < 0 && error("Order of derivative must be non-negative")
+    order == 0 && return p
+    hasnan(p) && return Polynomial(T[NaN], p.var)
+    order > length(p) && return zero(Polynomial{T})
+
+    n = length(p)
+    a2 = Vector{T}(undef, n - order)
+    @inbounds for i in order:n - 1
+        a2[i - order + 1] = reduce(*, (i - order + 1):i, init = p[i])
+    end
+    return Polynomial(a2, p.var)
+end
+
+function companion(p::Polynomial{T}) where T
+    d = length(p) - 1
+    d < 1 && error("Series must have degree greater than 1")
+    d == 1 && return diagm(0 => [-p[0] / p[1]])
+
+    R = eltype(one(T) / p.coeffs[end])
+    comp = diagm(-1 => ones(R, d - 1))
+    monics = p.coeffs ./ p.coeffs[end]
+    comp[:, end] .= -monics[1:d]
+    return comp
+end
+
+function Base.:+(p1::Polynomial, p2::Polynomial)
+    p1.var != p2.var && error("Polynomials must have same variable")
+    n = max(length(p1), length(p2))
```

Comment:
In #188 it is seen that this should be n-1, or we could use `degree` in place of `length`